runtime/compiler/optimizer/TreeLowering.cpp

```diff
@@ -38,13 +38,401 @@ TR::TreeLowering::optDetailString() const throw()
+      //   +--------------------------------+
+      //   | ttprev                         |
+      //   | ArrayStoreCHK                  |
+      //   |   astorei/awrtbari             |
+      //   |     aladd                      |
+      //   |       <array-reference>        |
+      //   |       index-offset-calculation |
+      //   |     <value-reference>          |
+      //   +--------------------------------+
+      //
+      // into
+      //   +--------------------------------+
+      //   | treetop                        |
+      //   |   <array-reference>            |
+      //   | treetop                        |
+      //   |   <value-reference>            |
+      //   | ificmpeq  -->------------------*---------+
+      //   |   iand                         |         |
+      //   |     iloadi <isClassFlags>      |         |
+      //   |       aloadi <componentClass>  |         |
+      //   |         aloadi <vft-symbol>    |         |
+      //   |           <array-reference>    |         |
+      //   |     iconst J9ClassIsValueType  |         |
+      //   |   iconst 0                     |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //   | BBStart (Extension)            |         |
+      //   | NULLCHK                        |         |
+      //   |   Passthrough                  |         |
+      //   |     <value-reference>          |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //                   |                          |
+      //                   +--------------------------+
+      //                   |
+      //                   v
+      //   +--------------------------------+
+      //   | BBStart                        |
+      //   | ArrayStoreCHK                  |
+      //   |   astorei/awrtbari             |
+      //   |     aladd                      |
+      //   |       aload <array>            |
+      //   |       index-offset-calculation |
+      //   |     aload <value>              |
+      //   +--------------------------------+
+      //
+      TR::SymbolReference *vftSymRef = comp()->getSymRefTab()->findOrCreateVftSymbolRef();
+      TR::SymbolReference *arrayCompSymRef = comp()->getSymRefTab()->findOrCreateArrayComponentTypeSymbolRef();
+      TR::SymbolReference *classFlagsSymRef = comp()->getSymRefTab()->findOrCreateClassFlagsSymbolRef();
+
+      TR::Node *vft = TR::Node::createWithSymRef(node, TR::aloadi, 1, anchoredArrayTT->getNode()->getFirstChild(), vftSymRef);
+      TR::Node *arrayCompClass = TR::Node::createWithSymRef(node, TR::aloadi, 1, vft, arrayCompSymRef);
+      TR::Node *loadClassFlags = TR::Node::createWithSymRef(node, TR::iloadi, 1, arrayCompClass, classFlagsSymRef);
+      TR::Node *isValueTypeNode = TR::Node::create(node, TR::iand, 2, loadClassFlags, TR::Node::iconst(node, J9ClassIsValueType));
```

Comment:
Is it better to move the creation of this node to frontend/j9 somewhere (e.g. VMJ9.cpp) since it directly refers to `J9ClassIsValueType` in the optimizer (that is meant to avoid referring to J9-isms overtly) ?

Reply:
As this part of Leo's change is simply moving code from `J9CodeGenerator.cpp` to `TreeLowering.cpp` without any changes, I suggest that we not let this comment hold up the addition of the new optimization.  If you're OK with it, I will address this in a follow on pull request.

Reply:
Thanks for the recommendation.  I hadn't been aware of that practice of avoiding that sort of dependency on J9 in the optimizer.  May I ask you to look at pull request #12126 for this change?